README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

<p align="center">

<img  src="https://i.ibb.co/Snv5grs/deepdivelogo.png">

</p>

[![](https://www.r-pkg.org/badges/version/deepdive?color=green)](https://cran.r-project.org/package=deepdive)
[![](https://img.shields.io/badge/Dev-1.0.2-green.svg)](https://rajeshb24.github.io/deepdive/)

# deepdive

**deepnet-\>deeptree-\>deepforest**

**Github version has been updated with minor bug fixes.Will be upated in CRAN soon.**

This package aims to provide simple intuitive functions to create quick
prototypes of artificial neural network or deep learning models for
general purpose application. In addition, check out experimental
algorithms from my personal research , **deeptree** and **deepforest**
for special cases to achieve better accuracy / generalization.

**deeptree**: This algorithm builds a CART tree to divide the solution
space in to leaves and fits an artificial neural network to each leaf.
This approach takes advantage of distinct properties of a tree and
neural network. It has tendency to overfit but multiple parameters can
be tuned to achieve better generalization. This model has a provision to
stack predictions from other models(currently stacking is only available
for regression).

**deepforest**: This algorithm builds multiple deepnets/deeptrees from
which either best deepnet can be selected by passing all variable and
data to each network or random deepnets/deeptrees based on random cuts
of variable/data can be combined together over a error choice.

*Reach me Out* : <rajeshbalakrishnan24@gmail.com> for any suggestions
and doubts or you can always leave a comment on github.

## Installation

You can install released version from CRAN or development from github
deepdive from [GitHub](https://github.com/RajeshB24/deepdive) with:

``` r

#CRAN

install.packages("deepdive")

#Development Version
devtools::install_github("RajeshB24/deepdive")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(deepdive)
 x <- data.frame(a = runif(1000)*100,
 b = runif(1000)*200,
 c = runif(1000)*100
 )
 y<- data.frame(y=20*x$a +30* x$b+10*x$c +10)

 #Training increase iterations for convergence
 modelnet<-deepnet(x,y,c(2,2),
 activation = c('relu',"sin"),
 reluLeak = 0.001,
 modelType = "regress",
 iterations =20,
 eta=0.8,
 optimiser="adam")
```

    ## iteration 3: 3411.16051267071

    ## iteration 7: 2604.11288801726

    ## iteration 11: 1937.08720004024

    ## iteration 15: 1989.74302007208

    ## iteration 20: 2087.62656554103

``` r
 #predict
# predDeepNet<-predict.deepnet(modelnet,newData=x)

 #evaluate
#sqrt(mean((predDeepNet$pred_y-y$y)^2))
```


NEWS.md
#### deepdive (1.0.2)

* deepdive::predict.deeptree() fixed mismatched prediction rows when data is sampled.(#2)
* deepdive::predict.deepnet() fixed issue which error on newData with singlerow.(#5)


R/deepInverseNets.R


#' @title Build and train an Artificial Neural Network of any size
#' @description Build and train Artifical Neural Network of any depth in a single line code. Choose the hyperparameters to improve the accuracy or generalisation of model.
#' @param x  a data frame with input variables
#' @param y  a data frame with ouptut variable
#' @param hiddenLayerUnits a numeric vector, length of vector indicates number of hidden layers and each element in vector indicates corresponding hidden units Eg: c(6,4) for two layers, one with 6 hiiden units and other with 4 hidden units. Note: Output layer is automatically created.
#' @param activation one of "sigmoid","relu","sin","cos","none". The default is "sigmoid". Choose a activation per hidden layer
#' @param reluLeak  numeric. Applicable when activation is "relu". Specify value between 0 any number close to zero below 1. Eg: 0.01,0.001 etc
#' @param modelType one of "regress","binary","multiClass". "regress" for regression will create a linear single unit output layer. "binary" will create a single unit sigmoid activated layer. "multiClass" will create layer with units corresponding to number of output classes with softmax activation.
#' @param iterations integer. This indicates number of iteratios or epochs in backpropagtion .The default value is 500.
#' @param eta numeric.Hyperparameter,sets the Learning rate for backpropagation. Eta determines the convergence ability and speed of convergence.
#' @param seed numeric. Set seed with this parameter. Incase of sin activation sometimes changing seed can yeild better results. Default is 2
#' @param gradientClip numeric. Hyperparameter numeric value which limits gradient size for weight update operation in backpropagation. Default is 0.8 . It can take any postive value.
#' @param regularisePar numeric. L2 Regularisation Parameter .
#' @param optimiser one of "gradientDescent","momentum","rmsProp","adam". Default value "adam"
#' @param parMomentum numeric. Applicable for optimiser "mometum" and "adam"
#' @param inputSizeImpact numeric. Adjusts the gradient size by factor of percentage of rows in input. For very small data set setting this to 0 could yeild faster result. Default is 1.
#' @param parRmsPropZeroAdjust numeric. Applicable for optimiser "rmsProp" and "adam"
#' @param parRmsProp numeric.Applicable for optimiser "rmsProp" and "adam"
#' @param printItrSize numeric. Number of iterations after which progress message should be shown. Default value 100 and for iterations below 100 atleast 5 messages will be seen
#' @param showProgress logical. True will show progress and F will not show progress
#' @param stopError  Numeric. Rmse at which iterations can be stopped. Default is 0.01, can be set as NA in case all iterations needs to run.
#' @param miniBatchSize integer. Set the mini batch size for mini batch gradient
#' @param useBatchProgress logical. Applicable for miniBatch , setting T will use show rmse in Batch and F will show error on full dataset. For large dataset set T
#' @param ignoreNAerror logical. Set T if iteration needs to be stopped when predictions become NA
#'
#'
#' @return returns model object which can be passed into \code{\link{predict.deepnet}}
#' @export
#'@importFrom fastDummies dummy_cols
#'@importFrom graphics barplot
#'@importFrom stats formula predict runif
#' @examples
#' \dontrun{
#' x <- data.frame(a = runif(1000)*100,
#' b = runif(1000)*200,
#' c = runif(1000)*100
#' )
#' y<- data.frame(y=20*x$a +30* x$b+10*x$c +10)
#'
#' #train
#' modelnet<-deepnet(x,y,c(2,2),
#' activation = c('relu',"sin"),
#' reluLeak = 0,
#' modelType = "regress",
#' iterations =4000,
#' eta=0.8,
#' optimiser="adam")
#'
#' #predict
#' predDeepNet<-predict.deepnet(modelnet,newData=x)
#'
#' #evaluate
#'sqrt(mean((predDeepNet$pred_y-y$y)^2))
#'
#' }
#'
#'
deepinversenet <- function(x,
                           y,
                           hiddenLayerUnits = c(2, 2),
                           activation = c('sigmoid', "sigmoid"),
                           reluLeak = 0,
                           modelType = c('regress'),
                           iterations = 500,
                           eta = 10 ^ -2,
                           seed = 2,
                           gradientClip = 0.8,
                           regularisePar = 0,
                           optimiser = "adam",
                           parMomentum = 0.9,
                           inputSizeImpact = 1,
                           parRmsPropZeroAdjust = 10 ^ -8,
                           parRmsProp = 0.9999,
                           printItrSize = 100,
                           showProgress = T,
                           stopError = 0.01,
                           miniBatchSize = NA,
                           useBatchProgress = T,
                           ignoreNAerror = F) {



  reluLeak <-ifelse(reluLeak==0,
                    0.001,
                    reluLeak)
  if (is.na(miniBatchSize)) {
    miniBatchSize = nrow(x)
  }


  set.seed(seed)

  xcolnames <- names(x)
  if (is.character(y)) {
    y <- as.factor(y)
  }

  if (modelType == "multiClass") {
    y <- dummy_cols(y)[, -1]

  }


  xType <- sapply(x, class)

  xfctrChrColsIdx <- which(xType %in% c("character", "factor"))
  if (length(xfctrChrColsIdx) > 0L) {
    x <- dummy_cols(x)
    x <- x[, -xfctrChrColsIdx]
  }


  inputColCount = ncol(x)
  outputColCount = ncol(y)

  #number of variables columns input in each layer
  interVariableCount <- c(inputColCount + 1, hiddenLayerUnits + 1)


  #nummber hiddenunits in each layer
  interOutCount <- c(hiddenLayerUnits, outputColCount)


  weightMatrix <<-
    lapply(c(1:length(interVariableCount)),
           function(i) {
             invweightInitialiser(i, interVariableCount, interOutCount, seed*i, activation)
           })







  previousWeightUpdate <-  lapply(c(1:length(interVariableCount)),
                                  function(i) {
                                    weightUpdateInitialiser(i, interVariableCount, interOutCount, seed)
                                  })





  previousWeightAdapt <-  lapply(c(1:length(interVariableCount)),
                                 function(i) {
                                   weightUpdateInitialiser(i, interVariableCount, interOutCount, seed)
                                 })


  printItrSize <- max(min(round(iterations / 5, 0), printItrSize),1)
  msgIter <- seq(0, iterations, by = printItrSize)

  if (max(msgIter) < iterations) {
    msgIter <- c(msgIter, iterations)
  }

  msgIter<-c(1:iterations)
####################################### Iterations ##################################################

costList<-c()
  prevCost<-NA

  for (itr in 1:iterations) {


    if (itr == 1) {
      prevBatch = 0
    }

    miniBatchIndex <- miniBatchCreate(nrow(x), miniBatchSize, prevBatch)
    prevBatch <- miniBatchIndex$bacthNo

    batchupper = miniBatchIndex[, "batchupper"]
    batchlower = miniBatchIndex[, "batchlower"]

    AllWeights <- invbackProp(
      as.matrix(x[batchlower:batchupper, ]),
      as.matrix(y[batchlower:batchupper, ]),
      weightMatrix,
      activation,
      reluLeak,
      modelType,
      eta,
      gradientClip,
      regularisePar,
      itr,
      optimiser,
      parMomentum,
      parRmsProp,
      parRmsPropZeroAdjust,
      previousWeightUpdate,
      previousWeightAdapt,
      inputSizeImpact
    )



    weightMatrix <-  AllWeights$weightMatrix

    previousWeightUpdate    <- AllWeights$previousWeightUpdate
    previousWeightAdapt    <- AllWeights$previousWeightAdapt

    if (itr %in% msgIter) {
      if (useBatchProgress == T) {
        itry <- as.matrix(y[batchlower:batchupper, ])
        feedList <-
          invfeedForward(as.matrix(cbind(x[batchlower:batchupper, ],1)),
                      weightMatrix,
                      activation,
                      reluLeak,
                      modelType)
      } else{
        itry <- as.matrix(y)
        feedList <- invfeedForward(as.matrix(cbind(x,1)),
                                weightMatrix,
                                activation,
                                reluLeak,
                                modelType)
      }



      feedOut <- feedList$a_output
      itrypred <- feedOut[[length(feedOut)]]



      #rmse

      costFun <- sqrt(mean((itrypred - itry) ^ 2))

       if(is.na(prevCost)){
         prevCost=costFun
       }

      if(costFun<prevCost){

        finalPred=itrypred
      finalWeights<-weightMatrix
      }

      costList[length(costList)+1]<-costFun

    }

    if (showProgress == T) {
      if (itr %in% msgIter) {
        print(paste0(
          "iteration ",
          ifelse(itr == iterations, itr, itr - 1)
          ,
          ": ",
          costFun
        ))
      }


    }
    if (itr %in% msgIter) {
      if (is.na(costFun)) {
        if (ignoreNAerror == F) {
          print("NA error please change eta")

          break()
          ;}
      }
      if (is.nan(costFun)) {
        print("NaN error please change eta")
        break()
        ;
      }


      if (!is.na(stopError) & !is.na(costFun)) {
        if (costFun <= stopError) {
          print(paste0("iteration ", itr, ": ", costFun))
          print("Reached stop error")
          break()
          ;

        }
      }

    }



  }


plot(costList,type="line")

  deepnetmod <- list(weightMatrix = finalWeights,
    activation = activation,
    modelType = modelType,
    outColMax = outColMax,
    outColMin = outColMin,
    inColMax = inColMax,
    inColMin = inColMin,
    reluLeak = reluLeak,
    xcolnames = xcolnames)

  class(deepnetmod) <- "deepnet"

  return(deepnetmod)

}



